### DISCUSSION TITLE
DrawString and DrawText and DrawFont

### DISCUSSION THREAD

**Discussion by piotrek1668** [amerkoleci/Vortice.Windows#303] [General]
Is there a way to draw a text/font/string with Direct3D9, Direct3D11 or Direct3D12 in Vortice.Windows?

In SlimDX I could create a Font and call DrawString on the font object

**Comment by Axiverse** [comment]
You can use DirectWrite directly in D3D11 or in D3D12 with D3D11On12. Font in SlimDX is not part of the d3d library, but instead the d3dx helper library which is not part of the core API. D3D12 does not otherwise have these helper libraries and you will have to manage it.

https://github.com/Chlumsky/msdfgen is a popular way for rendering fonts from libraries, but you will have to write your own libraries to do so. It does pretty well in my program:

![image](https://user-images.githubusercontent.com/535918/172338840-fdee0008-6fb5-4706-9287-7b4ad4d0e2b9.png)

**Reply by piotrek1668** [reply]
OK, thank you for your answer. So there is no way to draw text/font with Vortice.Direct3D9?

**Comment by Axiverse** [comment]
SlimDX has it ported, so it can be possible. Would probably require a contribution to the codebase https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX.Direct3D9/Mapping.xml#L2491 Possibly in a separate repo/project as it's in a helper library.

I'm not sure D3D9 would be high on other's priority lists, so if it's something you'd like to tackle that'd probably be the fastest approach.

**Reply by Axiverse** [reply]
Also if you're just using d3d9, you can probably just stick to sharpdx. I think it should still work. These libraries were forward looking - filling in new things where sharpdx stopped. So new features in d3d12, dxr, directml, etc. which aren't in SharpDX.

**Reply by piotrek1668** [reply]
Thank you for your answers. I switched to D3D11 and now draw the Text/Font via D2D on an IDXGISurface.
